# Sesión de ejercicios 8

El objetivo de esta sesión de ejercicios es que el estudiante aprenda a determinar cuando implementar una función y como invocarla correctamente. Adicionalmente, el estudiante debe aplicar la documentación adecuada desde el punto de vista del usuario y programador. Finalmente, que el estudiante cree y aplique funciones a problemas reales para de esta forma implementar código más legible y menos propenso a errores.

En esta sesión también los estudiantes tendrán la oportunidad de probar la funcionalidad del Sense Hat mediante un emulador en línea. 

# Ejercicio 1

El módulo Sense HAT es una placa electrónica que se adiciona al Raspberry Pi. Tiene una pantalla RGB LED de 8x8 elementos, un joystick de 5 botones, e inclue los siguientes sensores:

- Giroscopio
- Acelerómetro
- Magnetómetro
- Temperatura
- Presión barométrica
- Humedad

Ir a la siguiente dirección: https://projects.raspberrypi.org/en/projects/getting-started-with-the-sense-hat y leer detenidamente el tutorial desde el inicio hasta el punto `Setting orientation`. 

<img src="img1.PNG" style="height:300px;">

Durante el tutorial, se le pedirá que cree una cuenta en el sitio del emulador en línea: https://trinket.io/sense-hat

<img src="img2.png"  style="height:300px;">

### Tarea

Luego de haber leído detenidamente y probado el código del tutorial. Crear un programa que se ejecute indefinidamente y vaya mostrando en la pantalla flechas de distinto color que irán cambiando de dirección de forma aleatoria cada dos segundos.

Nota: las flechas pueden ser horizontales y verticales únicamente.  Si bien no se puede probar el código desde esta notebook, el estudiante deberá incluir el código de su solución en la celda de abajo.

Nota: el programa debe funcionar correctamente y el día de la próxima sesión de ejercicios los grupos deberán mostrar el programa funcionando

In [ ]:
# inserte su código aquí



# Ejercicio 2

Escribir y probar una función que genere un patrón similar a:

`*` <br>
`* *` <br>
`* * *` <br>
`* * * *` <br>
`* * * * *` <br>
`* * * *` <br>
`* * *` <br>
`* *` <br>
`*` <br>

Para generar este patrón, la función tendrá un parámetro `profundidad` (para el ejemplo anterior, `profundidad=9`). Adicionalmente este parámetro deberá ser impar. 

In [2]:
def get_patron(profundidad):
    """Construye un patrón de '*' con una cantidad de lineas igual a la profundidad.
    
    Parametros
    ----------
    profundidad : int
      cantidad de lineas que tendra el patrón
      
    Retorna
    -------
    patron : str
      patrón generado por la función
    """
    patron = ''
    linea = ''
    if profundidad % 2 != 0:
        for i in range(profundidad):
            if i > profundidad // 2:
                linea = linea[:-2]
            else:
                linea += '* '
            patron += f'\n{linea}'
    return patron


def main():
    profundidad = int(input('\nIngrese un número entero positivo impar: '))
    
    print(f'{get_patron(profundidad)}\n')


if __name__ == '__main__':
    main()


Ingrese un número entero positivo impar: 9

* 
* * 
* * * 
* * * * 
* * * * * 
* * * * 
* * * 
* * 
* 



# Ejercicio 3

Cree y pruebe una función que dada una lista (`numeros`) de números enteros y un número entero (`x`), retorne 3 valores: (1) cuantos números iguales `x` existen; (2) cuantos números menores a `x` existen; y, (3) cuantos números mayores a `x` existen.

Ejemplo:

numeros: [10, 5, 4, 4, 6, 3, 7, 10] <br>
x: 4

salida esperada:
Existen  2  números iguales a  4
Existen  5  números menores a  4
Existen  1  números mayores a  4

In [3]:
def get_mayores_menores_iguales(numeros, x):
    """Calcula la cantidad de numeros mayores, menores o iguales a 'x' existen en la lista 'numeros'.
    
    Parametros
    ----------
    numeros : int[]
      lista con numeros enteros
    
    x : int
      Un número entero para hacer la comparación
      
    Retorna
    -------
    iguales : int
      cantidad de números iguales a 'x'
      
    menores : int
      cantidad de números menores a 'x'
      
    mayores : int
      cantidad de números mayores a 'x'
    """
    iguales, menores, mayores = 0, 0, 0
    for i in numeros:
        if x < i:
            mayores += 1
        elif x > i:
            menores += 1
        else:
            iguales += 1
    return iguales, menores, mayores


def main():
    # Lista de prueba para el ejercicio
    numeros = [10, 5, 4, 4, 6, 3, 7, 10]
    
    x = int(input('\nIngrese un número: '))
    
    iguales, menores, mayores = get_mayores_menores_iguales(numeros, x)
    print(f'\nExisten {mayores} números mayores, {menores} menores y {iguales} iguales a {x}.\n')   


if __name__ == '__main__':
    main()


Ingrese un número: 4

Existen 5 números mayores, 1 menores y 2 iguales a 4.



# Ejercicio 4

Tomando como base el ejercicio 10 de la sesión de ejercicios 5, crear una función que dado un string, que representa una cédula, retorne True si la cédula es válida, caso contrario que retorne False

Nota: Básese en su propia resolución del ejercicio y optimice el código lo más posible, ayudándose de slices, list comprehensions, etc.

In [4]:
def get_validadores(cedula):
    """Calcula un número entero resultante de sumar todos los digitos de la 'cedula' con la condicion de que
       todas las posiciones impares se multiplequen por 2 y si el resultado de esta ultima operacion es un 
       número mayor o igual a 10, le restamos 9. Este número sirve para validar una cédula.
    
    Parametros
    ----------
    cedula : str
      Un texto que contiene los 10 dígitos de una cédula
      
    Retorna
    -------
    int(str(validador)[-1]) : int
      Último dígito del resultado de la operacion anterior que sirve para validar la cédula.
      
    digitos[-1] : int
      Último dígito de la cédula, necesario para la validación.
    """
    # Generamos una lista transformando a entero todos los dígitos de la cédula
    digitos = [int(i) for i in cedula]
    validador = 0
    for i in range(len(digitos) - 1):
        resultante = digitos[i]
        if i % 2 == 0:
            resultante *= 2
        validador += resultante if resultante < 10 else resultante - 9
    return int(str(validador)[-1]), digitos[-1]


def validar_cedula(cedula):
    """Valida que la 'cedula' sea correcta haciendo uso del último dígito para validar, siempre y cuando
       la cédula posea 10 dígitos.
    
    Parametros
    ----------
    cedula : str
      Un texto que contiene los 10 dígitos de una cédula.
      
    Retorna
    -------
    True o False : bool
      True si la cédula es correcta caso contrario False.
    """
    if cedula.isdigit() and len(cedula) == 10:
        digt_validador, ultm_digito = get_validadores(cedula)
        if ultm_digito == 0 and digt_validador == 0:
            return True
        if ultm_digito == 10 - digt_validador: 
            return True
    return False


def main():
    cedula = input('\nIngrese un número de cédula: ')
    if validar_cedula(cedula):
        print(f'\nEl número de cédula "{cedula}" es correcto\n')
    else:
        print(f'\nEl número de cédula "{cedula}" no es correcto!\n')


if __name__ == '__main__':
    main()


Ingrese un número de cédula: 0105736326

El número de cédula "0105736326" es correcto



# Ejercicio 5

Las funciones pueden verse también como expresiones, esto implica que se puede asociar una referencia de una función a una variable y usarla como una función. Esta funcionalidad se utiliza cuando se necesitan evaluar varias funciones con los mismos parámetros. 

Analice el ejemplo a continuación:

In [5]:
def funcion1(x):
    """Implementación de una función cuadrática
    
    Parametros
    ----------
    x : int
      valor de x de entrada
      
    Retorna
    -------
    x**2 : int
      resultado de la función
    """
    return x**2


def funcion2(x):
    """Implementación de una función cúbica

    Parametros
    ----------
    x : int
      valor de x de entrada
      
    Retorna
    -------
    x**3 : int
      resultado de la función
    """
    return x**3


# funciones a ser ejecutadas
funciones = [funcion1, funcion2]

# valores de x a ser probados
x_vals = range(0, 10)

for f in funciones: # se recorren las funciones una a una
    res = []  # acumulador de resultados
    
    # calcula los valores de la función de acuerdo a la función y valor de x actuales
    for x in x_vals:
        res.append(f(x))
    
    # muestra el resultado
    print(res)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 8, 27, 64, 125, 216, 343, 512, 729]


Modifique el código anterior para que se implemente una nueva función `funcion3` que ejecute la siguiente función: $(x-5)^2 - 2x + 3$. Se definirán valores de $x$ desde $1..30$ y la ejecución será como sigue: primero se ejecutará `funcion1`, luego `funcion2`, y finalmente `funcion3`. `funcion1` evaluará los valores de $x=\{1, 4, 7, 10, ...\}$. `funcion2` evaluará los valores $x=\{2, 5, 8, 11, ...\}$ y `funcion3` evaluará los valores $x=\{3, 6, 9, 12, ...\}$. El programa mostrará los valores de las funciones para los números números evaluados.

Salida esperada:

[1, 16, 49, 100, 169, 256, 361, 484, 625, 784] <br>
[8, 125, 512, 1331, 2744, 4913, 8000, 12167, 17576, 24389] <br>
[1, -8, 1, 28, 73, 136, 217, 316, 433, 568]

In [6]:
def funcion3(x):
    """Implentación de función matemática.

    Parametros
    ----------
    x : int
      valor de x de entrada.
      
    Retorna
    -------
    (x - 5)**2 - 2*x + 3 : int
      resultado de operación matemática.
    """
    return (x - 5)**2 - 2*x + 3


def main():
    funciones = [funcion1, funcion2, funcion3]
    limite_inicio = 1 # Primer número con el que trabajará la primera función
    salida = f'\nSalida:\n\n'

    for f in funciones:
        # Genera una lista con los resulsultados de las operaciones de cada función.
        res = [f(i) for i in range(limite_inicio, 31, 3)]
        salida += f'{res}\n'
        limite_inicio += 1 # Número de inicio con el que trabajará la 'n' función.

    print(f'{salida}\n')


if __name__ == '__main__':
    main()


Salida:

[1, 16, 49, 100, 169, 256, 361, 484, 625, 784]
[8, 125, 512, 1331, 2744, 4913, 8000, 12167, 17576, 24389]
[1, -8, 1, 28, 73, 136, 217, 316, 433, 568]




# Ejercicio 6

Modifique el código del ejercico anterior para optimizarlo de forma tal que no sea necesario realizar iteraciones sobre `x` : `for x in x_vals:`.

Nota: utilice la función `map`. 

In [7]:
def main():
    funciones = [funcion1, funcion2, funcion3]
    limite_inicio = 1 # Número de inicio con el que trabajará la primera función
    salida = f'\nSalida:\n\n'

    for f in funciones:
        # Genera una lista con los resulsultados de las operaciones de cada función.
        res = list(map(f, range(limite_inicio, 31, 3)))
        salida += f'{res}\n'
        limite_inicio += 1 # Número de inicio con el que trabajará la 'n' función.

    print(f'{salida}\n')


if __name__ == '__main__':
    main()


Salida:

[1, 16, 49, 100, 169, 256, 361, 484, 625, 784]
[8, 125, 512, 1331, 2744, 4913, 8000, 12167, 17576, 24389]
[1, -8, 1, 28, 73, 136, 217, 316, 433, 568]




# Ejercicio 7

Modifique el ejercicio anterior para que, en vez de usar definciones de funciones tradicionales i.e. `funcion1`, `funcion2`, y `funcion3`, utilice funciones lambda para implementar la misma funcionalidad

In [8]:
# Calcula el cuadrado de un número 'x'
funcion1 = lambda x: x**2


# Calcula el cubo de un número 'x'
funcion2 = lambda x: x**3


# Calcula el valor de una función matemática con el valor de un número 'x'
funcion3 = lambda x: (x - 5)**2 - 2*x + 3


# Genera una lista con los resultados de pasar una lista de números por cualquiera de las anteriores funciones.
clasificar_por_fn = lambda fn, rango: f'\n{list(map(fn, rango))}'


def main():
    funciones = [funcion1, funcion2, funcion3]
    limite_inicio = 1 # Número de inicio con el que trabajará la primera función
    salida = f'\nSalida:\n'

    for f in funciones:
        salida += clasificar_por_fn(f, range(limite_inicio, 31, 3))
        limite_inicio += 1 # Número de inicio con el que trabajará la 'n' función.

    print(f'{salida}\n')


if __name__ == '__main__':
    main()


Salida:

[1, 16, 49, 100, 169, 256, 361, 484, 625, 784]
[8, 125, 512, 1331, 2744, 4913, 8000, 12167, 17576, 24389]
[1, -8, 1, 28, 73, 136, 217, 316, 433, 568]



# Ejercicio 8

Implemente un programa que dada una lista con mensajes codificados, verificar cuantas veces se repite un patrón en ellos. Para ello, primeramente deberá generar dichos mensajes, i.e. una lista de `n` mensajes donde cada elemento es un string de `m` caracteres randómicos que van desde el caracter `a` hasta `k`. Una lista de mensajes válida tiene la siguiente forma: 

['ekjeg', 'hgkch', 'biaae', 'fcfhb', 'akcbg', 'jhdaj', 'geibk', 'ebchj', 'heeae', 'kfeid'] <br>

Para $n=10$ y $m=5$

Luego, buscar un patrón cualquiera de 2 caractéres en cada uno de los elementos de esta lista. El programa deberá retornar cuantas veces encontró el patrón en la lista. Se contabilizan únicamente los elementos que contenga el patrón una sola vez.

Ejemplo: <br>
mensaje: ['keeab', 'dhabe', 'afebd', 'cabab', 'fkabi', 'kifca', 'bfiai', 'hhjbk', 'jkjij', 'echgf'] <br>
patron: ab

salida esperada: 3

Nota: para generar valores aleatorios dentro de un conjunto, se puede usar la función choice de la librería random

In [11]:
import random

# Genera un texto con caracteres aleatorios de la lista pasada por parametro, con la longitud indicada.
get_text = lambda lista, longitud: ''.join([random.choice(lista) for i in range(longitud)])


def get_num_ocurrencias(texto, patron):
    """Calcula el número de veces que se repite un patron dentro de un texto.

    Parametros
    ----------
    texto : str
      Un texto cualquira dentro del cual se desea buscar un patrón.
    
    patron : str
      Patŕon que se desea buscar dentro del texto.
      
    Retorna
    -------
    ocurrencias : int
      Número de veces que se repitio el patrón dentro del texto.
    """
    limt_inicial = 0
    limt_final = len(patron)
    ocurrencias = 0
    while limt_final < len(texto):
        if texto[limt_inicial:limt_final] == patron:
            ocurrencias += 1
        limt_inicial += 1
        limt_final += 1
    return ocurrencias


def main():
    # Lista con caracteres de prueba
    letras = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k']
    
    n = random.randint(10, 30) # Cantidad de texto que deseamos generar.
    m = random.randint(5, 15) # Cantidad de caracteres por texto.

    # Generamos una lista con textos aleatoros.
    mensajes = [get_text(letras, m) for i in range(n)]
    
    # Generamo un texto patrón con caracteres aleatorios
    patron = get_text(letras, 2)

    ocurrencias = 0
    for i in mensajes:
        # Contamos unicamente lo textos en los que el patrón aparece solo una vez.
        if get_num_ocurrencias(i, patron) == 1:
            ocurrencias += 1

    print(f'\nLista: {mensajes}\n\nEl patrón: "{patron}" aparece {ocurrencias} veces en la lista.\n') 


if __name__ == '__main__':
    main()


Lista: ['ieffjgba', 'abeadkeg', 'bkkbagdf', 'djbghgcj', 'khkfabjj', 'cdgbaake', 'egadbajk', 'bbhgdbbh', 'gagcddga', 'ifaegfgd', 'fecfjfdk', 'ikkjchjk', 'ckiedecj', 'echghghf', 'bdfjgfhh', 'fageghba', 'ecbjibch', 'eaahiikf', 'kkbdchkj']

El patrón: "cd" aparece 2 veces en la lista.



# Ejercicio 1 exámen interciclo

Se requiere escribir un programa que dada una lista de listas conteniendo números enteros, genere una lista de una sola dimensión contenido todos los números. Esta operación comúnmente se conoce como aplanar una lista. Esta lista además, deberá ser ordenada de forma descendente. Por ejemplo: <br><br>
Entrada: [[1, 12, 3], [100, 13, 22, 0], [15, 30]] <br>
Salida: [100, 30, 22, 15, 13, 12, 3, 1, 0]

Para solucionar este ejercicio, el alumno no podrá hacer uso de ninguna función preestablecida en Python.

In [1]:
# Genera una lista con los elementos de cada una de las listas contenidas en la lista pasada por parametro.
aplanar_lista = lambda listas: [y for i in listas for y in i]


def ordenar_descendente(lista):
    """Ordena de forma descendente los elementos de una lista de números enteros.

    Parametros
    ----------
    lista : int[]
      Un lista con números enteros.
    
    Retorna
    -------
    lista : int[]
      lista con sus elementos ordenados de forma descendente.
    """
    # Comparamos cada elemento de la lista con lo demas elementos
    for i in range(len(lista)):
         # Enviamos hacia el final de la lista a los elementos menores.
        cont = 0
        while cont < len(lista) - 1:
            # Si el número es menor intercambiamos sus posiciones.
            if lista[cont] < lista[cont + 1]:
                aux = lista[cont]
                lista[cont] = lista[cont + 1]
                lista[cont + 1] = aux
            cont += 1
    return lista


def main():
    # Lista con número enteros de prueba.
    entrada = [[1, 12, 3], [100, 13, 22, 0], [15, 30]]
    salida = ordenar_descendente(aplanar_lista(entrada))
    print(f'\nSalida: {salida}\n')


if __name__ == '__main__':
    main()


Salida: [100, 30, 22, 15, 13, 12, 3, 1, 0]



# Ejercicio 2 exámen interciclo

Se requiere crear un programa para encriptar un mensaje. La lógica es la siguiente: el mensaje original será transformado moviendo las posiciones de los caracteres en base a la lista numérica provista. Para este ejercicio, se le debe pedir al usuario una `frase` y el `pad`. El pad podrá ser cualquier número de 3 cifras. Por ejemplo: <br><br>
Frase: el zorro es una pelicula vieja <br>
pad: 123 <br>
salida: fn aqusq fu vpd rhmkfvnd xlfld <br>

Analizando la salida: <br>
e -> f movido una posición <br>
l -> n movido dos posiciones <br>
“ “ -> los caracteres vacíos no se transforman  <br>
z -> a movido una posición<br>
o -> q movido dos posiciones<br>
r -> u movido tres posiciones

Entonces, pad es una secuencia con números que se reutiliza para el proceso. Además, las posiciones del alfabeto pueden verse como circulares; es decir, si se debe mover z dos posiciones, el resultado será b
Ayuda: El siguiente código genera una lista con los caracteres de la a-z:

<code>list(map(chr, range(97, 123)))  # [‘a’, ‘b’, ‘c’, …, ‘x’, ‘y’, ‘z’] </code>

In [3]:
def get_char_encript(letra, num_recorrer):
    """Cambia una letra por otra que se encuantra en una posición que resulta de sumar 
       'num_recorrer' a su posición.

    Parametros
    ----------
    letra : char
      Letra que deseamos cambiar.
    
    num_recorrer : int
      Número que se debe sumar a la posicion de la letra actual para obtener una nueva letra.
    
    Retorna
    -------
    letra_encr : char
      Retorna una núeva letra siempre y cuando la recibida por parametro no ea un texto vacio.
    """
    abecedario = list(map(chr, range(97, 123)))
    letra_encr = letra
    for index, value in enumerate(abecedario):
        if letra == value:
            next_posc = index + num_recorrer
            num_items = len(abecedario)
            if next_posc > num_items - 1:
                letra_encr = abecedario[next_posc - num_items]
                break
            else:
                letra_encr = abecedario[next_posc]
                break
    return letra_encr


def get_encript_text(texto, path):
    """Encripta un texto cambiando cada letra por otra que esta en una posición superior del abecedario,
       para esto utiliza una lista con la cantidades enteras para el cambio de posición

    Parametros
    ----------
    texto : str
      Texto que se desea encriptar.
    
    path : int[]
      Lista de números que servirán para encriptar el texto.
    
    Retorna
    -------
    encript_text : str
      Texto encriptado.
    """
    encript_text = ""
    # Para cada letra se utilizara un número del path
    cont = 0
    for i in texto:
        if cont == len(path):
            cont = 0
        if cont < len(path):
            encript_text += get_char_encript(i, path[cont])
        cont += 1
    return encript_text


def main():
    frase = input('\nIngrese una frase: ')
    path = input('Ingrese un path para encriptar, debe ser un numero entero de 3 cifras: ')

    frase_encriptada = get_encript_text(frase, [int(i) for i in path])

    print(f'\nLa frase encriptada es: {frase_encriptada}.\n')


if __name__ == '__main__':
    main()


Ingrese una frase: el zorro es una pelicula vieja
Ingrese un path para encriptar, debe ser un numero entero de 3 cifras: 123

La frase encriptada es: fn aqusq fu vpd rhmkfvnd xlfld.

